src/build_system.ml

```diff
@@ -999,61 +1006,88 @@ The following targets are not:
+
+  add_rules_exn t targets_here;
+
   remove_old_artifacts t ~dir ~subdirs_to_keep;
 
-  List.iter alias_rules ~f:(compile_rule t);
-  Option.iter alias_dir ~f:(fun alias_dir ->
-    remove_old_artifacts t ~dir:alias_dir ~subdirs_to_keep);
+  let alias_targets =
+    (match alias_rules with
+     | None ->
+       None
+     | Some f ->
+       Some (f ~subdirs_to_keep))
+  in
+  Loaded.Build {
+    rules_produced;
+    targets_here;
+    targets_of_alias_dir =
+      match alias_targets with
+      | None ->
+        Path.Map.empty
+      | Some v ->
+        v
+  }
 
-  { rules_produced; targets }
 
 let get_rule_other t fn =
   let dir = Path.parent_exn fn in
   if Path.is_in_build_dir dir then
-    load_dir t ~dir;
-  Fiber.return (Path.Table.find t.files fn)
+    (match load_dir t ~dir with
+     | Non_build _ -> assert false
+     | Build { targets_here; _ } ->
+       (Path.Map.find targets_here fn))
+  else
+    None
 
 and get_rule t path =
-  match Path.Table.find t.files path with
-  | Some _ as some -> Fiber.return some
-  | None ->
-    let dir = Path.parent_exn path in
-    if Path.is_strict_descendant_of_build_dir dir then begin
-      load_dir t ~dir;
-      match Path.Table.find t.files path with
-      | Some _ as some -> Fiber.return some
-      | None ->
-        let loc = Rule_fn.loc () in
-        no_rule_found t ~loc path
-    end else if Path.exists path then
-      Fiber.return None
-    else
+  let dir = Path.parent_exn path in
+  if Path.is_strict_descendant_of_build_dir dir then begin
+    let rules = load_dir_and_get_buildable_targets t ~dir in
+    match Path.Map.find rules path with
+    | Some _ as some -> Fiber.return some
+    | None ->
       let loc = Rule_fn.loc () in
-      Errors.fail_opt loc
-        "File unavailable: %s" (Path.to_string_maybe_quoted path)
+      no_rule_found t ~loc path
+  end else if Path.exists path then
+    Fiber.return None
+  else
+    let loc = Rule_fn.loc () in
+    Errors.fail_opt loc
+      "File unavailable: %s" (Path.to_string_maybe_quoted path)
 
 let all_targets () =
   let t = t () in
-  String.Map.iter t.contexts ~f:(fun ctx ->
-    File_tree.fold t.file_tree ~traverse_ignored_dirs:true ~init:()
-      ~f:(fun dir () ->
-        load_dir t
-          ~dir:(Path.append_source ctx.Context.build_dir (File_tree.Dir.path dir))));
-  Path.Table.foldi t.files ~init:[] ~f:(fun key _ acc -> key :: acc)
+  String.Map.to_list t.contexts
+  |> List.concat_map ~f:(fun (_, ctx) ->
+    File_tree.fold t.file_tree ~traverse_ignored_dirs:true ~init:[]
+      ~f:(fun dir acc ->
+        match
+          load_dir t
+            ~dir:(
+              Path.append_source ctx.Context.build_dir
+                (File_tree.Dir.path dir))
+        with
+        | Non_build _ -> acc
+        | Build {
+          targets_here;
+          targets_of_alias_dir;
+          _
+        } ->
+          Path.Map.keys targets_of_alias_dir
+          @ Path.Map.keys targets_here @ acc
```

Comment:
Let's just have `all_targets` return `Path.Set.t`. It's more clear and gets rid of this slowness.

Reply:
Pretty sure adding to a set (`O(log(n))`) is slower than adding to a list (`O(1)`). 

I'm making this change because I don't want the order in this list to matter, but it's almost certainly going to make things slower.

Reply:
I was under the impression that appending lists is `O(n)`. Do we know we are adding a small list?

In any case it doesn't matter so much. Keeping the order stable is much more important 👍

Reply:
Just to clarify:

Appending `a @ acc` is `O(length a)` (note that the length of `acc` does not matter).

Insertion of a list into a set, on the other hand:
```
fold_left a ~init:acc ~f:Set.add
```
 is `O(length a * log(Set.length acc + length a))`

`length(a)` is a shared factor between the two, so the difference is the extra log-factor in the case of a set.